api/file.go

```diff
@@ -224,27 +184,132 @@ func (h FileHandler) Delete(ctx *gin.Context) {
 		_ = ctx.Error(err)
 		return
 	}
-	err = os.Remove(m.Path)
+	err = h.delete(ctx, m)
 	if err != nil {
-		if !os.IsNotExist(err) {
-			_ = ctx.Error(err)
+		_ = ctx.Error(err)
+		return
+	}
+
+	h.Status(ctx, http.StatusNoContent)
+}
+
+// create a file.
+func (h FileHandler) create(ctx *gin.Context, name string) (m *model.File, err error) {
+	mode := ctx.ContentType()
+	switch mode {
+	case binding.MIMEMultipartPOSTForm:
+		m, err = h.createMultipart(ctx, name)
+	default:
+		m, err = h.createBody(ctx, name)
+	}
+	return
+}
+
+// create a file with multipart form.
+func (h FileHandler) createMultipart(ctx *gin.Context, name string) (m *model.File, err error) {
+	input, err := ctx.FormFile(FileField)
+	if err != nil {
+		err = &BadRequestError{err.Error()}
+		return
+	}
+	m = &model.File{}
+	m.Name = name
+	m.Encoding = input.Header.Get(ContentType)
+	m.CreateUser = h.BaseHandler.CurrentUser(ctx)
+	db := h.DB(ctx)
+	err = db.Create(&m).Error
+	if err != nil {
+		return
+	}
+	defer func() {
+		if err != nil {
+			h.Status(ctx, http.StatusInternalServerError)
+			_ = db.Delete(&m)
 			return
 		}
+	}()
+	reader, err := input.Open()
+	if err != nil {
+		err = &BadRequestError{err.Error()}
+		return
 	}
-	err = h.DB(ctx).Delete(m).Error
+	defer func() {
+		_ = reader.Close()
+	}()
+	writer, err := os.Create(m.Path)
 	if err != nil {
-		_ = ctx.Error(err)
 		return
 	}
+	defer func() {
+		_ = writer.Close()
+	}()
+	_, err = io.Copy(writer, reader)
+	if err != nil {
+		return
+	}
+	err = os.Chmod(m.Path, 0666)
+	if err != nil {
+		return
+	}
+	return
+}
 
-	h.Status(ctx, http.StatusNoContent)
+// create a file with request body.
+func (h FileHandler) createBody(ctx *gin.Context, name string) (m *model.File, err error) {
+	m = &model.File{}
+	m.Name = name
+	m.Encoding = binding.MIMEYAML
```

Comment:
**hint**: hard coded to YAML is fine since json is a subset.  Tested and it worked fine.